backend/pkg/console/schema_registry.go

```diff
@@ -0,0 +1,569 @@
+	})
+
+	return result, nil
+}
+
+// SchemaRegistrySubjectDetails represents a schema registry subject along
+// with other information such as the registered versions that belong to it,
+// or the full schema information that's part of the subject.
+type SchemaRegistrySubjectDetails struct {
+	Name                string                                `json:"name"`
+	Type                schema.SchemaType                     `json:"type"`
+	Compatibility       schema.CompatibilityLevel             `json:"compatibility"`
+	RegisteredVersions  []SchemaRegistrySubjectDetailsVersion `json:"versions"`
+	LatestActiveVersion int                                   `json:"latestActiveVersion"`
+	Schemas             []*SchemaRegistryVersionedSchema      `json:"schemas"`
+}
+
+const (
+	// SchemaVersionsAll can be specified as version to retrieve all schema versions.
+	SchemaVersionsAll string = "all"
+	// SchemaVersionsLatest can be specified as version to retrieve the latest active schema.
+	SchemaVersionsLatest string = "latest"
+)
+
+// GetSchemaRegistrySubjectDetails retrieves the schema details for the given subject, version tuple.
+// Use version = 'latest' to retrieve the latest schema.
+// Use version = 'all' to retrieve all schemas for this subject.
+// Use version = 3 to retrieve the specific version for the given subject
+func (s *Service) GetSchemaRegistrySubjectDetails(ctx context.Context, subjectName, version string) (*SchemaRegistrySubjectDetails, error) {
+	// 1. Retrieve all schema versions registered for the given subject
+	versions, err := s.getSchemaRegistrySchemaVersions(ctx, subjectName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve subject versions: %w", err)
+	}
+	latestActiveVersion := -1
+	softDeletedVersions := make(map[int]bool)
+	for _, v := range versions {
+		if v.IsSoftDeleted {
+			softDeletedVersions[v.Version] = true
+			continue
+		}
+		if v.Version > latestActiveVersion {
+			latestActiveVersion = v.Version
+		}
+	}
+
+	// 2. Retrieve schemas and compat level concurrently
+	var compatLevel schema.CompatibilityLevel
+
+	grp, grpCtx := errgroup.WithContext(ctx)
+
+	grp.Go(func() error {
+		// 2. Retrieve subject config (subject compatibility level)
+		configRes, err := s.kafkaSvc.SchemaService.GetSubjectConfig(grpCtx, subjectName)
+		if err != nil {
+			s.logger.Warn("failed to get subject config", zap.String("subject", subjectName), zap.Error(err))
+			return nil
+		}
+		compatLevel = configRes.Compatibility
+		return nil
+	})
+
+	// If version 'all' request schemas for all versions individually
+	versionsToRetrieve := []string{version}
+	if version == SchemaVersionsAll {
+		versionsToRetrieve = make([]string, len(versions))
+		for i, v := range versions {
+			versionsToRetrieve[i] = strconv.Itoa(v.Version)
+		}
+	}
+
+	schemasCh := make(chan *SchemaRegistryVersionedSchema, len(versionsToRetrieve))
+	for _, v := range versionsToRetrieve {
+		copiedVersion := v
+		grp.Go(func() error {
```

Comment:
Should we bound this concurrency? Probably unlikely to be an issue, but if there are a lot of versions it could result in a lot of concurrent requests to the SR API.

Reply:
I wondered too! We are not really sure what rate limits we should apply though because the SR has never been benchmarked - the majority of the retrieved information seems to be served from memory. I think the larger factor that drives the concurrency here is the number of concurrent incoming requests (one request for each sr subject).

I think it's very rare to have a significant number of versions (e.g. more than 20). However, I added a limit of 10 concurrent Go routines, it shouldn't hurt as you said